gsy_framework/sim_results/carbon_emissions/results.py

```diff
@@ -113,149 +36,127 @@ def _find_start_and_end_dates(
-                    )
-                    df_carbon_ratio = pd.concat([df_carbon_ratio, df_daily])
-        elif (
             country_code in MONTHLY_CARBON_EMISSIONS_COUNTRY_CODES
         ):  # source: https://www.electricitymaps.com/data-portal
+
             file_path = (
                 f"gsy_framework/resources/carbon_ratio_per_country/{country_code}_2023_monthly.csv"
             )
-            df_file = pd.read_csv(file_path, index_col=0, parse_dates=True).tz_localize("UTC")
-            full_range = pd.date_range(start=start, end=end, freq="H", tz="UTC")
-            df_carbon_ratio = df_file[["Carbon Intensity gCO₂eq/kWh (direct)"]].reindex(
-                full_range, method="ffill"
-            )
-            df_carbon_ratio.rename(
-                columns={"Carbon Intensity gCO₂eq/kWh (direct)": "Ratio (gCO2eq/kWh)"},
-                inplace=True,
-            )
+            data = {
+                datetime.strptime(row["Datetime (UTC)"], "%Y-%m-%d %H:%M:%S").replace(
+                    tzinfo=UTC
+                ): float(row["Carbon Intensity gCO₂eq/kWh (direct)"])
+                for row in csv.DictReader(open(file_path, mode="r"))
+            }
+
+            full_range = self._full_range(start, end)
+            carbon_ratio = [
+                {
+                    "time": time,
+                    "Ratio (gCO2eq/kWh)": data.get(
+                        time, data[max(k for k in data if k.month == time.month)]
+                    ),
+                    "country_code": country_code,
+                }
+                for time in full_range
+            ]
+
         else:  # source: https://ourworldindata.org/grapher/carbon-intensity-electricity
             country_code_alpha3 = countries.get(country_code).alpha3
             file_path = (
                 "gsy_framework/resources/carbon_ratio_per_country"
                 "/carbon-intensity-electricity_yearly.csv"
             )
-            df_file = pd.read_csv(file_path, index_col=2)
-            df_file = df_file[df_file["Code"] == country_code_alpha3]
-
-            df_file = df_file.loc[df_file.index.get_level_values(0).max()]  # filter fix max year
-            full_range = pd.date_range(start=start, end=end, freq="H", tz="UTC")
-            df_carbon_ratio = pd.DataFrame(index=full_range)
-            df_carbon_ratio["Ratio (gCO2eq/kWh)"] = df_file[
-                "Carbon intensity of electricity - gCO2/kWh"
+
+            # Find the carbon ratio for the most recent year
+            with open(file_path, mode="r") as f:
+                data = [
+                    (int(row["Year"]), float(row["Carbon intensity of electricity - gCO2/kWh"]))
+                    for row in csv.DictReader(f)
+                    if row["Code"] == country_code_alpha3
+                ]
+            max_year = max(year for year, _ in data)
+            carbon_ratio = next(value for year, value in data if year == max_year)
```

Comment:
NIT. I know it works, but I personally would prefer creating a dict and just use a lookup rather than adding two more loops. I know that these are not expensive here, but maybe we can discuss whether my following proposal would be better or worse.
```suggestion
            with open(file_path, mode="r") as f:
                data = {
                    int(row["Year"]): float(row["Carbon intensity of electricity - gCO2/kWh"])
                    for row in csv.DictReader(f)
                    if row["Code"] == country_code_alpha3
                }
            carbon_ratio = data[max(data.keys())]
```

Reply:
Thanks, the dict approach performs better.